README.md
# antigen.garnish 2

Human and mouse ensemble tumor neoantigen prediction from SNVs and complex variants. Immunogenicity filtering based on the [Tumor Neoantigen Selection Alliance (TESLA)](https://www.parkerici.org/research-project/tumor-neoantigen-selection-alliance-tesla/).

## Citation

> Richman LP, Vonderheide RH, and Rech AJ. Neoantigen dissimilarity to the self-proteome predicts immunogenicity and response to immune checkpoint blockade. Cell Systems. 2019.

## Selected references

> Duan, F., Duitama, J., Seesi, S.A., Ayres, C.M., Corcelli, S.A., Pawashe, A.P., Blanchard, T., McMahon, D., Sidney, J., Sette, A., et al. Genomic and bioinformatic profiling of mutational neoepitopes reveals new rules to predict anticancer immunogenicity. J Exp Med. 2014.

> Luksza, M, Riaz, N, Makarov, V, Balachandran VP, et al. A neoepitope fitness model predicts tumour response to checkpoint blockade immunotherapy. Nature. 2017.

> Rech AJ, Balli D, Mantero A, Ishwaran H, Nathanson KL, Stanger BZ, Vonderheide RH. Tumor immunity and survival as a function of alternative neopeptides in human cancer. Clinical Cancer Research, 2018.

> Wells DK, van Buuren MM, Dang KK, Hubbard-Lucey VM, Sheehan KCF, Campbell KM, Lamb A, Ward JP, Sidney J, Blazquez AB, Rech AJ, Zaretsky JM, Comin-Anduix B, Ng AHC, Chour W, Yu TV, Rizvi1 H, Chen JM, Manning P, Steiner GM, Doan XC, The TESLA Consortium, Merghoub T, Guinney J, Kolom A, Selinsky C, Ribas A, Hellmann MD, Hacohen N, Sette A, Heath JR, Bhardwaj N, Ramsdell F, Schreiber RD, Schumacher TN, Kvistborg P, Defranoux N. Key Parameters of Tumor Epitope Immunogenicity Revealed Through a Consortium Approach Improve Neoantigen Prediction. Cell. 2020.

## Installation

Two methods exist to run `antigen.garnish`:

1. Docker
2. Linux

### Docker

```sh
docker pull andrewrech/antigen.garnish:2.3.1

cID=$(docker run -it -d andrewrech/antigen.garnish:2.3.1 /bin/bash)
```

[Download](https://services.healthtech.dtu.dk/software.php) netMHC binaries (academic license): NetMHC 4.0, NetMHCpan 4.1b, NetMHCII 2.3, NetMHCIIpan 4.0.

Copy netMHC `tar.gz` files to the container and run the installation script:

```sh
docker cp netMHC-4.0a.Linux.tar.gz $cID:/netMHC-4.0a.Linux.tar.gz
docker cp netMHCII-2.3.Linux.tar.gz $cID:/netMHCII-2.3.Linux.tar.gz
docker cp netMHCpan-4.1b.Linux.tar.gz $cID:netMHCpan-4.1b.Linux.tar.gz
docker cp netMHCIIpan-4.0.Linux.tar.gz $cID:netMHCIIpan-4.0.Linux.tar.gz

docker exec $cID config_netMHC.sh
```

### Linux

#### Dependencies

- R &ge; 3.5.0
- [Bioconductor](https://www.bioconductor.org/install/) [Biostrings package](https://www.bioconductor.org/packages/release/bioc/html/Biostrings.html)
- [GNU Parallel](https://www.gnu.org/software/parallel/)
- [mhcflurry](https://github.com/openvax/mhcflurry)
- netMHC tools (see below)
- tcsh (required for netMHC, install via your package manager)

#### Installation

Install the dependencies listed above. Then, download and extract `antigen.garnish` data:

```sh
ANTIGEN_GARNISH_DIR="~/antigen.garnish"

cd ~
curl -fsSL "https://s3.amazonaws.com/get.rech.io/antigen.garnish-2.3.0.tar.gz" | tar -xvz
chmod 700 -R "$ANTIGEN_GARNISH_DIR"
```

Install antigen.garnish:

```r
# install.packages("remotes")
remotes::install_github("andrewrech/antigen.garnish")
```

Next, [download](https://services.healthtech.dtu.dk/software.php) netMHC binaries (academic license): NetMHC 4.0, NetMHCpan 4.1b, NetMHCII 2.3, NetMHCIIpan 4.0.

Move the binaries into the `antigen.garnish` data directory, first setting the `NET_MHC_DIR` and `ANTIGEN_GARNISH_DIR` environment variables:

```sh
NET_MHC_DIR=/path/to/folder/containing/netMHC/downloads

cd "$NET_MHC_DIR"
mkdir -p "$ANTIGEN_GARNISH_DIR/netMHC"

tar xvzf netMHC-4.0a.Linux.tar.gz -C "$ANTIGEN_GARNISH_DIR/netMHC"
tar xvzf netMHCII-2.3.Linux.tar.gz -C "$ANTIGEN_GARNISH_DIR/netMHC"
tar xvzf netMHCpan-4.1b.Linux.tar.gz -C "$ANTIGEN_GARNISH_DIR/netMHC"
tar xvzf netMHCIIpan-4.0.Linux.tar.gz -C "$ANTIGEN_GARNISH_DIR/netMHC"

chown "$USER" "$ANTIGEN_GARNISH_DIR/netMHC"
chmod 700 -R "$ANTIGEN_GARNISH_DIR/netMHC"
```

## Usage

See the [reference manual](https://s3.amazonaws.com/get.rech.io/antigen.garnish.pdf).

### Docker

#### Interactive use

Copy GRCh38-annotated VCF files and/or metadata including HLA alleles onto the running container using the `docker cp` [command](https://docs.docker.com/engine/reference/commandline/cp/). The container ID is still saved as `$cID` from the installation above. You will also need to use this command and container ID to copy saved output files from the docker container after you complete your analysis.

Copy any needed files onto the running container, for example:

```sh

docker cp myfile.txt $cID:/myfilecopy.txt

```

Now launch the interactive virtual machine with the container you started:

```sh

docker exec -it $cID bash
R

```

```r

library(antigen.garnish)

```

Follow the instructions in the next section titled **Linux** to complete your interactive R analysis. When you complete your analysis, copy any desired output files off the container to your local machine with the `docker cp` [command](https://docs.docker.com/engine/reference/commandline/cp/). Shut down and clean up your container like this:

```sh

docker cp $cID:/myoutput.txt ~/myagdockeroutput.txt

docker stop $cID

docker rm $cID

```

### Linux

Parallel cores used can be set via environment variable AG_THREADS (default: all available).

#### Predict neoantigens from missense mutations, insertions, and deletions

```r
library(magrittr)
library(data.table)
library(antigen.garnish)

# load an example VCF
dir <- system.file(package = "antigen.garnish") %>%
       file.path(., "extdata/testdata")

file <- file.path(dir, "TUMOR.vcf")

# extract variants
dt <-  garnish_variants(file)

# add space separated MHC types
# see list_mhc() for nomenclature of supported alleles
# MHC may also be set to "all_human" or "all_mouse" to use all supported alleles

dt[, MHC := c("HLA-A*01:47 HLA-A*02:01 HLA-DRB1*14:67")]

# predict neoantigens
result <- dt %>% garnish_affinity(.)

result %>% str
```

#### Predict neoantigens from Microsoft Excel or other table input

Transcript ID level input table format:
```
# sample_id ensembl_transcript_id cDNA_change MHC
# sample_1  ENST00000311936       c.718T>A    HLA-A*02:01 HLA-A*03:01
# sample_1  ENST00000311936       c.718T>A    H-2-Kb H-2-Kb
```

Protein level input (with optional WT paired input) table format:
```
# sample_id pep_mut           pep_wt            mutant_index MHC
# sample_1  MTEYKLVVVDADGVGK  MTEYKLVVVDAGGVGK  12           HLA-A*02:01
# sample_1  MTEYKLVVVDDDGVGK  MTEYKLVVVDAGGVGK  12 13        HLA-A*02:01
# sample_1  MTEYKLVVVDAGGAAA  MTEYKLVVVDAGGVGK  14 15 16     HLA-A*02:01
# sample_1  SIINFEKLMILKATFI  MTEYKLVVVDAGGVGK  all          HLA-A*02:01
```

```r
library(magrittr)
library(data.table)
library(antigen.garnish)
library(rio) # package to import Excel and other tables

# load an example table
dir <- system.file(package = "antigen.garnish") %>%
       file.path(., "extdata/testdata")

file <- file.path(dir, "antigen.garnish_example_peptide_with_WT_input.txt")

# read in excel or other format file with rio::import and convert to data table
# or substitute the path to your file here
mytable <- rio::import(file) %>% data.table::as.data.table()

# only use first two rows of table for example
mytable <- mytable[1:2]

# predict neoantigens from data table object
result <- garnish_affinity(mytable)

result %>% str
```

#### Directly calculate foreignness score and dissimilarity for a list of sequences

```r
library(magrittr)
library(data.table)
library(antigen.garnish)

# generate our character vector of sequences
v <- c("SIINFEKL", "ILAKFLHWL", "GILGFVFTL")

# calculate foreignness score
v %>% foreignness_score(db = "human") %>% print

# calculate dissimilarity
v %>% dissimilarity_score(db = "human") %>% print
```

#### How are peptides generated?

```r
library(magrittr)
library(data.table)
library(antigen.garnish)

data.table::data.table(
   pep_base = "Y___*___THIS_IS_________*___A_PEPTIDE_TEST!______*__X",
   mutant_index = c(5, 25, 47, 50),
   pep_type = "test",
   var_uuid = c(
                "front_truncate",
                "middle",
                "back_truncate",
                "end")) %>%
   make_nmers %>% print
```

## Acknowledgments

We thank the follow individuals for contributions and helpful discussion:

- [David Balli](https://www.linkedin.com/in/davidballi1)
- [Adham Bear](https://www.med.upenn.edu/apps/faculty/index.php/g20001100/p1073)
- [Katelyn Byrne](https://www.parkerici.org/person/katelyn-byrne/)
- [Danny Wells](http://dannykwells.com/)

## License

Please see [LICENSE](https://github.com/immune-health/antigen.garnish/blob/master/LICENSE).


NEWS.md
# antigen.garnish 2.3.1

Bugfix:

- errors parsing rare MHCII alleles for netMHC tools

# antigen.garnish 2.3.0

Bugfix:

- update NetMHC tool data directory paths
- fix permissions on configuration script (was non-executable)
- minor documentation improvements

# antigen.garnish 2.2.0

Bugfix:

- remove netMHC binary check from foreignness_score function that does not require netMHC binaries

# antigen.garnish 2.1.0

Added functionality:

- add Dockerfile used to build publically available Docker image (andrewrech/antigen.garnish)
- relax parsing of SnpEff annotations to reduce errors across VCFs and/or SnpEff versions
- simplify internal functions that configure netMHC tools
- use Docker to run tests

# antigen.garnish 2.0.0

Breaking changes:

- new filtering criteria based on TESLA: https://www.parkerici.org/research-project/tumor-neoantigen-selection-alliance-tesla/, see ?garnish_antigens
- update Ensembl transcript CDS database to release 101 (any prior versioned transcript release is acceptable)
- remove support for mitochondrial CDS, xlsx input, JAFFA input, PureCN input to focus on peptide/transcript tables and VCFs
- improve SnpEff annotation parsing

Additional updates:

- allow setting thread use via environment variable AG_THREADS
- include percentile rank filtering at default netMHC thresholds
- simplify installation instructions
- remove 7 dependencies
- additional integration tests
- do not export internal functions

# antigen.garnish 1.1.1

- update to include configuration for netMHC dependencies
- support for paired wild-type and mutant protein level input

# antigen.garnish 1.1.0

- added dissimilarity_score
- updated readme citation, link to manuscript repository
- externalized smith-waterman alignment function
- appropriate test updates
- remove deprecated garnish_fitness function
- update to LICENSE

# antigen.garnish 1.0.0

- prepare for the release of the antigen.garnish manuscript
- multiple metrics of antigen quality are now computed
- dissimilarity_score dummy function present
- pulled most unnecessary mclapply loops to prevent parallelization failures
- functionalized foreignness_score code into external function
- garnish_plot, garnish_summary now incorporate antigen quality metrics
- garnish_antigens now returns ranked antigens by quality metrics
- numerous edge case, efficiency updates and bug fixes

# antigen.garnish 0.0.6

- clonality filter added to garnish_affinity
- RNA expression filter added to garnish_affinity
- fitness model implemented in R vs. Python
- garnish_antigens function added to rank neoantigens
- garnish_score function added to return sample level immune fitness summary
- updated summary and plot functions with better output
- added wiki with installation instructions
- numerous edge case bug fixes
- remove biomaRt dependencies

# antigen.garnish 0.0.5

- add ncbi-blast functionality to determine neoantigen near matches
- add ncbi-blast functionality to determine known IEDB matches
- implement antigen fitness model of Lukza et al. _Nature_ 2017

# antigen.garnish 0.0.4

- add summary plots
- improve test coverage

# antigen.garnish 0.0.3

- inverse match against global normal proteins
- improve testing, test formatting

# antigen.garnish 0.0.2

- add prediction deduplication to garnish_affinity
- add gene fusions as a source of neoantigens using garnish_jaffa
- stable API

# antigen.garnish 0.0.1

- initial version


R/garnish_plots.R
## ---- garnish_plot
#' Graph summary results using ggplot2.
#'
#' Plots ADN, CDN, priority, frameshift, and fusion derived neoepitopes for Class I and Class II MHC by sample.
#' See `garnish_summary` for more information on neoepitope classification by antigen.garnish.
#'
#' @param dt Data table. Prediction output from `garnish_predictions`. Either dt or jdt must be provided as input.
#' @param jdt Data table. Prediction output from `garnish_jaffa`` %>% `garnish_predictions`. Either dt or jdt must be provided as input.
#' @param save Optional: Should the plot be saved to the working directory as antigen.garnish_summary.pdf? Default is TRUE.
#'
#' @examples
#'\dontrun{
#'library(magrittr)
#'library(antigen.garnish)
#'
#'  # download an example VCF
#'    g <- "antigen.garnish_example.vcf" %T>%
#'    utils::download.file("http://get.rech.io/antigen.garnish_example.vcf", .) %>%
#'
#'  # extract variants
#'    antigen.garnish::garnish_variants %>%
#'
#'  # add test MHC types
#'      .[, MHC := c("HLA-A*02:01 HLA-DRB1*14:67",
#'                   "H-2-Kb H-2-IAd",
#'                  "HLA-A*01:47 HLA-DRB1*03:07")] %>%
#'
#'  # predict neoepitopes
#'    antigen.garnish::garnish_predictions %>%
#'
#'  # plot it
#'    antigen.garnish::garnish_plot
#'   
#'  # here is our ggplot object that can be further customized 
#'   class(g)
#'  
#'  # save the plot with a custom ggplot theme 
#'  g + my_ggplot_custom_theme_object %>% cowplot::ggsave("antigen.garnish_custom_themed_plot.pdf")
#'   
#'  # show location of where default plot was saved
#'   list.files(pattern = "summary\\.pdf", full.names = TRUE)   
#'   
#'}
#'
#' @return
#'
#' A list of ggplot2 objects and, if save = TRUE, saves each of these objects pdfs in the working directory.
#' The first plot is a column graph indicating the number of peptides in each sample classified as ADN, CDN, priority, frameshift-derived, and fusion-derived neoepitopes
#' (if any) for each sample, faceted by Class I vs Class II MHC. The threshold for fusion and frameshift derived neoepitopes is Consensus_scores < 1000nM.
#' See `garnish_summary`` documentation for further explanation of neoepitope classification of ADN, CDN, and priority.
#' 
#' The second plot and third plots are returned only if frameshift and/or fusion mutants (respectively) are present in the input table. These plot
#' the number of peptides per sample by MHC class I or II and binned by binding affinity (1000 - 500nM, 500 - 50nM, and < 50 nM). 
#'
#' @export garnish_plot
#'
#' @md

garnish_plot <- function(dt = NULL, jdt = NULL, save = TRUE){
 
  ag_gg_theme <-
    ggplot2::theme(line = ggplot2::element_line(colour = "#000000")) +
    ggplot2::theme(axis.line = ggplot2::element_line(color="black")) +
    ggplot2::theme(plot.title = ggplot2::element_text(size = ggplot2::rel(2*1.2),
                                                      colour = "#000000", lineheight = 0.9, face = "bold", vjust = 0)) +
    ggplot2::theme(axis.title.x = ggplot2::element_text(colour = "#000000",
                                                        size = ggplot2::rel(2*1.425), face = "bold")) +
    ggplot2::theme(axis.title.y = ggplot2::element_text(colour = "#000000",
                                                        size = ggplot2::rel(2*1.425), face = "bold", angle = 90, vjust = 0)) +
    ggplot2::theme(axis.text.x = ggplot2::element_text(colour = "#000000",
                                                       size = ggplot2::rel(2*1.425), face = "bold", angle = 30, hjust = 0.9, vjust = 0.92)) +
    ggplot2::theme(axis.text.y = ggplot2::element_text(colour = "#000000",
                                                       size = ggplot2::rel(2*1.425), face = "bold", hjust = 0.9, vjust = 0.92, angle = 30)) +
    ggplot2::theme(legend.text = ggplot2::element_text(colour = "#000000",
                                                       size = ggplot2::rel(2*1))) +
    ggplot2::theme(strip.background = ggplot2::element_rect(fill = "#eeeeee")) +
    ggplot2::theme(strip.text = ggplot2::element_text(face = "bold")) +
    ggplot2::theme(legend.key = ggplot2::element_blank()) +
    ggplot2::theme(legend.background = ggplot2::element_rect(fill = "transparent", colour = NA)) +
    ggplot2::theme(panel.grid = ggplot2::element_blank()) +
    ggplot2::theme(panel.background = ggplot2::element_rect(fill = "transparent", colour = NA)) +
    ggplot2::theme(plot.background = ggplot2::element_rect(fill = "transparent", colour = NA)) +
    ggplot2::theme(plot.margin=grid::unit(c(0, 0, 0, 0 ), "cm"))
  
  ag_colors <-   c("#ff80ab",
                   "#b388ff",
                   "#82b1ff",
                   "#a7ffeb",
                   "#b9f6ca",
                   "#f4ff81",
                   "#ffe57f",
                   "#ff9e80",
                   "#ff8a80",
                   "#ea80fc",
                   "#8c9eff",
                   "#80d8ff",
                   "#84ffff",
                   "#ccff90",
                   "#ffff8d",
                   "#ffd180")
  
  ##check function input
  if (missing(dt) & missing(jdt)) stop("at least one dt from garnish_predictions must be provided.")
  
  if (!missing(dt) & !missing(jdt)) dt <- cat_tables(dt, jdt)
  
  if(!missing(dt)){
  
  if (!(c("nmer",
          "MHC",
          "sample_id",
          "DAI",
          "Consensus_scores") %chin% names(dt)) %>% any)
    stop("'sample_id', 'nmer', 'MHC', 'frameshift', 'DAI', and 'Consensus_scores' columns are required in dt")
  
  dt <- dt[pep_type != "wt"] %>% unique(by = c("nmer",
                                               "MHC",
                                               "sample_id",
                                               "DAI",
                                               "Consensus_scores"))
  
  dt <- dt %>% .[Consensus_scores < 5000] %>%
    .[pep_type == "mutnfs" & Consensus_scores < 50, type := "CDN"] %>%
      .[pep_type == "mutnfs" & DAI > 10, type := "ADN"] %>%
        .[pep_type == "mutnfs" & Consensus_scores < 50 & DAI > 10, type := "priority"] %>%
          .[pep_type == "fus" & Consensus_scores < 1000, type := "fusion"] %>%
            .[pep_type == "mut_other" & Consensus_scores < 1000, type := "frameshift"]
  
  if (nrow(dt) < 1) stop("No neoeptiopes with Consensus_scores < 5000nM ")
  
  dt_pl <- dt[MHC %like% "(HLA-[ABC]\\*)|(H-2-[A-Z][a-z])", MHC := "Class I"] %>%
    .[MHC %like% "(HLA-D[A-Z0-9]+\\*)|(H-2-[A-Z]{2}[a-z])", MHC := "Class II"] %>%
      .[!is.na(type)]
  
  if (nrow(dt_pl) < 1) stop("No neoepitopes meet classification criteria")
  
  gg_dt <- dt_pl[, .N, by = c("sample_id", "MHC", "type")]
  
  gdt <- dt_pl %>% (function(dt){
    
    ns <- dt[, sample_id %>% unique %>% length]
    
    gdt <- data.table(sample_id = dt[, sample_id %>% unique],
                      MHC = c(replicate(ns, "Class I"), replicate(ns, "Class II")),
                      type = c(replicate(ns * 2, "ADN"), replicate(ns * 2, "CDN"), replicate(ns * 2, "priority"),
                                replicate(ns * 2, "frameshift"), replicate(ns * 2, "fusion")),
                      N = 0) %>% unique
    return(gdt)
  })
  
  gg_dt <- merge(gg_dt, gdt, by = intersect(names(gg_dt), names(gdt)), all = TRUE) %>% 
    .[, N := max(N), by = c("sample_id", "type", "MHC")] %>% unique
  
  if (any(gg_dt[, sample_id %>% unique %>% nchar] > 7)){
    
    message("Sample_id has >7 characters, shortening names for aesthetics.
    To circumvent this, change sample_ids to less than 7 characters in the input data.table.")
    
    for (i in gg_dt[nchar(sample_id) > 7, sample_id %>% unique] %>% seq_along){
      
    gg_dt[sample_id == gg_dt[nchar(sample_id) > 7, sample_id %>% unique][i],
          sample_id := sample_id %>%
               substr(1, 7) %>% paste0(., "_", i)]
                                                                  }
                                                          }
  
 g1 <- ggplot2::ggplot(gg_dt, ggplot2::aes(x = sample_id, y = N)) +
    ggplot2::geom_col(ggplot2::aes(fill = type), col = "black", position = "dodge") +
    ggplot2::facet_wrap(~MHC) +
    ag_gg_theme +
    ggplot2::theme(legend.position = "bottom", legend.title = ggplot2::element_blank()) +
    ggplot2::scale_fill_manual(values = ag_colors) +
    ggplot2::theme(strip.text.x = ggplot2::element_text(size = ggplot2::rel(2))) +
    ggplot2::ylab("peptides") +
    ggplot2::xlab("") +
    ggplot2::ggtitle(paste0("antigen.garnish summary"))
 
 if(save == "TRUE") cowplot::ggsave(plot = g1, "antigen.garnish_summary.pdf", height = 6, width = 9)
  
 if (!(dt[, effect_type %>% unique] %like% "frameshift" %>% any)) g2 <- NA
 
 if (dt[, effect_type %>% unique] %like% "frameshift" %>% any){
 # frameshift plot
   
   dt_pl <- dt[MHC %like% "(HLA-[ABC]\\*)|(H-2-[A-Z][a-z])", MHC := "Class I"] %>%
     .[MHC %like% "(HLA-D[A-Z0-9]+\\*)|(H-2-[A-Z]{2}[a-z])", MHC := "Class II"] %>%
        .[effect_type %like% "frameshift" & Consensus_scores < 1000]
   
   dt_pl[, binding := "<1000nM"] %>% 
     .[Consensus_scores < 500, binding := "<500nM"] %>%
        .[Consensus_scores < 50, binding := "<50nM"]
   
gg_dt <- dt_pl[, .N, by = c("sample_id", "MHC", "binding")] 

gdt <- dt_pl %>% (function(dt){
  
  ns <- dt[, sample_id %>% unique %>% length]
 
   gdt <- data.table(sample_id = dt[, sample_id %>% unique],
                  MHC = c(replicate(ns, "Class I"), replicate(ns, "Class II")),
                  binding = c(replicate(ns * 2, "<50nM"), replicate(ns * 2, "<500nM"), replicate(ns * 2, "<1000nM")),
                  N = 0) %>% unique
        })

gg_dt <- merge(gg_dt, gdt, by = intersect(names(gg_dt), names(gdt)), all = TRUE) %>% 
  .[, N := max(N), by = c("sample_id", "binding", "MHC")] %>% unique
 
if (any(gg_dt[, sample_id %>% unique %>% nchar] > 7)){
  
  message("Sample_id has >7 characters, shortening names for aesthetics.
          To circumvent this, change sample_ids to less than 7 characters in the input data.table.")
  
  for (i in gg_dt[nchar(sample_id) > 7, sample_id %>% unique] %>% seq_along){
    
    gg_dt[sample_id == gg_dt[nchar(sample_id) > 7, sample_id %>% unique][i],
          sample_id := sample_id %>%
            substr(1, 7) %>% paste0(., "_", i)]
  }
}

 g2 <- ggplot2::ggplot(gg_dt, ggplot2::aes(x = sample_id, y = N)) +
   ggplot2::geom_col(ggplot2::aes(fill = binding), col = "black", position = "dodge") +
   ggplot2::facet_wrap(~MHC) +
   ag_gg_theme +
   ggplot2::theme(legend.position = "bottom", legend.title = ggplot2::element_blank()) +
   ggplot2::scale_fill_manual(values = ag_colors[1:3]) +
   ggplot2::theme(strip.text.x = ggplot2::element_text(size = ggplot2::rel(2))) +
   ggplot2::ylab("peptides") +
   ggplot2::xlab("") +
   ggplot2::ggtitle(paste0("Frameshift neoepitopes"))
 
 if (save == "TRUE") cowplot::ggsave(plot = g2, "antigen.garnish_Frameshifts_summary.pdf", height = 6, width = 9)
 
 }
  }
  
 # fusion plot
 
  if(!missing(jdt) & missing(dt)) jdt <- jdt
  
  if(!missing(jdt) & !missing(dt)) jdt <- dt
  
 if (!"fusion genes" %chin% (jdt %>% names)) g3 <- NA
 
 if ("fusion genes" %chin% (jdt %>% names)){
  
   
   dt_pl <- jdt[MHC %like% "(HLA-[ABC]\\*)|(H-2-[A-Z][a-z])", MHC := "Class I"] %>%
     .[MHC %like% "(HLA-D[A-Z0-9]+\\*)|(H-2-[A-Z]{2}[a-z])", MHC := "Class II"] %>%
     .[!is.na(fusion_uuid) & Consensus_scores < 1000]
   
   dt_pl[, binding := "<1000nM"] %>% 
     .[Consensus_scores < 500, binding := "<500nM"] %>%
     .[Consensus_scores < 50, binding := "<50nM"]
   
   gg_dt <- dt_pl[, .N, by = c("sample_id", "MHC", "binding")] 
   
   gdt <- dt_pl %>% (function(dt){
     
     ns <- dt[, sample_id %>% unique %>% length]
     
     gdt <- data.table(sample_id = dt[, sample_id %>% unique],
                       MHC = c(replicate(ns, "Class I"), replicate(ns, "Class II")),
                       binding = c(replicate(ns * 2, "<50nM"), replicate(ns * 2, "<500nM"), replicate(ns * 2, "<1000nM")),
                       N = 0) %>% unique
   })
   
   gg_dt <- merge(gg_dt, gdt, by = intersect(names(gg_dt), names(gdt)), all = TRUE) %>% 
     .[, N := max(N), by = c("sample_id", "binding", "MHC")] %>% unique
   
   if (any(gg_dt[, sample_id %>% unique %>% nchar] > 7)){
     
     message("Sample_id has >7 characters, shortening names for aesthetics.
             To circumvent this, change sample_ids to less than 7 characters in the input data.table.")
     
     for (i in gg_dt[nchar(sample_id) > 7, sample_id %>% unique] %>% seq_along){
       
       gg_dt[sample_id == gg_dt[nchar(sample_id) > 7, sample_id %>% unique][i],
             sample_id := sample_id %>%
               substr(1, 7) %>% paste0(., "_", i)]
     }
   }
   
   g3 <- ggplot2::ggplot(gg_dt, ggplot2::aes(x = sample_id, y = N)) +
     ggplot2::geom_col(ggplot2::aes(fill = binding), col = "black", position = "dodge") +
     ggplot2::facet_wrap(~MHC) +
     ag_gg_theme +
     ggplot2::theme(legend.position = "bottom", legend.title = ggplot2::element_blank()) +
     ggplot2::scale_fill_manual(values = ag_colors[1:3]) +
     ggplot2::theme(strip.text.x = ggplot2::element_text(size = ggplot2::rel(2))) +
     ggplot2::ylab("peptides") +
     ggplot2::xlab("") +
     ggplot2::ggtitle(paste0("Fusion neoepitopes"))
   
   if (save == "TRUE") cowplot::ggsave(plot = g3, "antigen.garnish_Fusions_summary.pdf", height = 6, width = 9)
   
 }
  
  if (missing(dt) & !missing(jdt)) return(g3)
  
  if (missing(jdt) & !missing(dt)) return(list(g1,g2))
 
 glist <- list(g1, g2, g3)
 
 return(glist)
 
}

 






## ---- cat_tables
#' Internal function to combine garnish_predictions outputs from garnish_jaffa
#' and garnish_variants or direct input to garnish_predictions.
#'
#' @export cat_tables
#' @md
#' 
cat_tables <- function(dt1, dt2){
  ##get the jaffa_dt
  if ("fus_tx" %chin% names(dt1) %>% any) jdt <- dt1
  if ("fus_tx" %chin% names(dt2) %>% any) jdt <- dt2
  
  ##get the vcf predictions output dt  
  if (grepl("mutnfs", (dt1[, .SD %>% unique, .SDcols = "pep_type"] %>%
                       unlist)) %>% any) vdt <- dt1
  if (grepl("mutnfs", (dt2[, .SD %>% unique, .SDcols = "pep_type"] %>%
                       unlist)) %>% any) vdt <- dt2
  
  ##subset tables for what we care about for downstream merge
  
  vdt <- vdt[, .SD, .SDcols = c("nmer", "sample_id", "MHC", "Consensus_scores", "DAI",
                                "Lower.CI", "Upper.CI", "ensembl_transcript_id",
                                "effect_type", "external_gene_name", "frameshift", "protein_change",
                                "cDNA_change", "mutant_index", "pep_mut", "pep_wt", "pep_type", "var_uuid", "dai_uuid", "nmer_uuid")]
  
  jdt <- jdt[, .SD, .SDcols = c("nmer", "sample_id", "MHC", "Consensus_scores", "Lower.CI", "Upper.CI",
                                "fusion genes", "frameshift", "mutant_index", "pep_mut", "pep_gene_1", "var_uuid",
                                "fusion_uuid", "nmer_uuid")]   
  
  ##add fus as pep_type for later analysis, add mutfs for frameshifts
  jdt[, pep_type := "fus"]
  
  vdt[is.na(pep_type) & effect_type %like% "frameshift",
      pep_type := "mutfs"]
  
  ##change names to overlap in the vcf dt
  strings <- jdt[, sample_id %>% unique]
  
  ###contruct our regex
  strings <- paste("(", strings, ")", sep = "", collapse = "|")
  
  ##rename stings in vcf dt from full BAM name
  vdt[, sample_id := sample_id %>% stringr::str_extract(pattern = strings)]
  
  ##final merge
  dto <- merge(vdt, jdt, all = TRUE, by = intersect(names(vdt), names(jdt)))
  return(dto)
}

  

